### PR TITLE
Fix ceres Jet issues with quaternion identity

### PR DESCRIPTION
**Pull request by sweeneychris**
This moves any invocation of sqrt(...) into guarded parts of code.

@strasdat please make sure what I'm doing is kosher here

### CODE REVIEW

**Comment by NikolausDemmel** [comment]
Related: https://github.com/strasdat/Sophus/pull/226

@sweeneychris, I'm just curious what exactly happens with Jet with the unpatched code for 0 angle? I would expect some nan or inf in the dual numbers, but since the derivative part is not used in the "if" branches, I would have expected that not to be an issue.

Minor comment: Maybe the checks should now use "epsilon squared" (just to leave the thresholds as they were; probably the changed thresholds are also ok...).

File: sophus/so3.hpp [hunk 1]
```diff
@@ -257,16 +256,18 @@ class SO3Base {
     /// Representation through Encapsulation of Manifolds"
     /// Information Fusion, 2011
 
-    if (n < Constants<Scalar>::epsilon()) {
-      /// If quaternion is normalized and n=0, then w should be 1;
-      /// w=0 should never happen here!
+    if (squared_n < Constants<Scalar>::epsilon()) {
```
Comment (strasdat) [inline sophus/so3.hpp]: I agree with @NikolausDemmel, lets keep the threshold where it is, e.g. also epsilon square, here and below.

File: sophus/so3.hpp [hunk 2]
```diff
@@ -257,16 +256,18 @@ class SO3Base {
...
+      // If quaternion is normalized and n=0, then w should be 1;
+      // w=0 should never happen here!
       SOPHUS_ENSURE(abs(w) >= Constants<Scalar>::epsilon(),
                     "Quaternion (%) should be normalized!",
                     unit_quaternion().coeffs().transpose());
       Scalar squared_w = w * w;
       two_atan_nbyw_by_n =
           Scalar(2) / w - Scalar(2) * (squared_n) / (w * squared_w);
+      J.theta = Scalar(0);
```
Comment (strasdat) [inline sophus/so3.hpp]: Setting theta=0 is a pretty crude approximation.

Maybe setting theta = 2 atan(squared_n/w), [hence](https://www.wolframalpha.com/input/?i=2+*+atan%28x+%2F+w%29+) 2*squared_n/w is better? Not sure.

Comment (strasdat) [reply sophus/so3.hpp]: Hm, this probably does not work any better. Maybe we want two if conditions?

if (squared_n< eps^4) {
  theta = 0
} else if (squared_n < eps ^2) {
  [taylor expansion]
} else {
  [analytic]
}

Comment (sweeneychris) [reply sophus/so3.hpp]: @strasdat  I'm not sure I follow why you'd want the eps^4 condition? The taylor series expansion seems appropriate and safe here

Comment (NikolausDemmel) [reply sophus/so3.hpp]: What's the idea behind `2 atan(squared_n/w)`? You can't expand it with `n` and just replace `n` with `squared_n` to still make it sepend somehow in the input? Or is there a deeper reason why replacing `n` with `squared_n` is suitable?

Also, I think `two_atan_nbyw_by_n` Taylor expansion in the line above should have 2/3 as constant for the "squared_n" term (it currently has just 2): https://www.wolframalpha.com/input/?i=2+*+atan%28x+%2F+w%29+

**Comment by NikolausDemmel** [comment]
> @sweeneychris, I'm just curious what exactly happens with Jet with the unpatched code for 0 angle? I would expect some nan or inf in the dual numbers, but since the derivative part is not used in the "if" branches, I would have expected that not to be an issue.

Do you have a quick comment on this? I should probably just test it myself...

**Pull request merged**